src/core/ds/atomicLazy.ml

```diff
@@ -0,0 +1,21 @@
+open Atomic
+
+type 'a t = {
+	mutable value: 'a option;
+	mutex: Mutex.t;
+	compute: unit->'a
+}
+
+let from_fun f =
+	{ value = None; mutex = Mutex.create (); compute = (fun () -> f()) }
+
+let force lazy_val =
+	match lazy_val.value with
+		| None ->
+			Mutex.protect lazy_val.mutex (fun () ->
```

Comment:
This is still not correct actually. Two threads could queue up here and then both end up executing `compute` one after the other, which is not what we want.

This might be a good use-case for a condition variable.